File: src/tools/getStackStatus.ts
import * as fs from "node:fs";
import { stackStateYamlPath } from "src/config";
import type { Tool, ToolProgress } from "src/Tool";
import type { ComposePsRow } from "src/services/docker/composeRunner";
import { scrubLine } from "src/state/secretRedactor";
import { z } from "zod";
import { collectSecretKeys } from "./shared/secretKeys";

export const GetStackStatusInputSchema = z.object({
  stackName: z.string(),
  tailLines: z.number().int().min(0).max(1000).optional(),
});

export type GetStackStatusInput = z.infer<typeof GetStackStatusInputSchema>;

export interface GetStackStatusResult {
  rows: ComposePsRow[];
  logTail: string;
}

export const getStackStatus: Tool<GetStackStatusInput, GetStackStatusResult> = {
  name: "get_stack_status",
  description: "Show container state, health, ports, and last log lines for a stack.",
  inputSchema: GetStackStatusInputSchema,
  category: "read-only",
  needsPermission: () => false,
  call: async function* (input, ctx): AsyncGenerator<ToolProgress, GetStackStatusResult> {
    const yamlPath = stackStateYamlPath(ctx.cwd, input.stackName);
    if (!fs.existsSync(yamlPath)) {
      return { rows: [], logTail: `stack ${input.stackName} not found` };
    }

    yield { type: "progress", msg: `Compose ps + logs for ${input.stackName}...` };
    const secretKeys = collectSecretKeys(input.stackName, ctx);
    const bound = ctx.composeRunner.forStack(input.stackName, yamlPath);
    const rows = await bound.ps({ json: true });
    const logs = bound.logs({ tailLines: input.tailLines ?? 50 });
    let logTail = "";

    while (true) {
      const r = await logs.next();
      if (r.done) break;
      logTail += scrubLine(r.value, secretKeys);
    }

    return { rows, logTail };
  },
};

File: python/src/docker_agent/tools/get_stack_status.py
"""get_stack_status tool.

Parity: ``src/tools/getStackStatus.ts``.
"""

from __future__ import annotations

from collections.abc import AsyncIterator
from pathlib import Path

from pydantic import BaseModel, ConfigDict, Field

from docker_agent.config import stack_state_yaml_path
from docker_agent.services.docker.compose_runner import ComposePsRow
from docker_agent.state.secret_redactor import scrub_line
from docker_agent.tool import ToolContext, ToolDone, ToolProgress
from docker_agent.tools.shared.secret_keys import SecretKeysContext, collect_secret_keys

_MODEL_CONFIG = ConfigDict(extra="forbid", populate_by_name=True)


class GetStackStatusInput(BaseModel):
    model_config = _MODEL_CONFIG

    stack_name: str = Field(alias="stackName")
    tail_lines: int | None = Field(default=None, alias="tailLines", ge=0, le=1000)


class GetStackStatusResult(BaseModel):
    model_config = _MODEL_CONFIG

    rows: list[ComposePsRow]
    log_tail: str = Field(alias="logTail")


class GetStackStatusTool:
    name = "get_stack_status"
    description = (
        "Show container state, health, ports, and last log lines for a stack."
    )
    input_schema = GetStackStatusInput
    category = "read-only"

    def needs_permission(self, _input: GetStackStatusInput) -> bool:
        return False

    async def call(
        self, input: GetStackStatusInput, ctx: ToolContext
    ) -> AsyncIterator[ToolProgress | ToolDone]:
        yaml_path = stack_state_yaml_path(input.stack_name, ctx.cwd)
        if not Path(yaml_path).exists():
            yield ToolDone(
                GetStackStatusResult(
                    rows=[],
                    logTail=f"stack {input.stack_name} not found",
                )
            )
            return

        yield ToolProgress(msg=f"Compose ps + logs for {input.stack_name}...")
        secret_keys = collect_secret_keys(
            input.stack_name,
            SecretKeysContext(cwd=ctx.cwd, state_store=ctx.state_store),
        )
        bound = ctx.compose_runner.for_stack(input.stack_name, yaml_path)
        rows = await bound.ps(json=True)
        log_tail = ""
        async for line in bound.logs(tail_lines=input.tail_lines or 50):
            log_tail += scrub_line(line, secret_keys)

        yield ToolDone(GetStackStatusResult(rows=rows, logTail=log_tail))


get_stack_status = GetStackStatusTool()

__all__ = [
    "GetStackStatusInput",
    "GetStackStatusResult",
    "GetStackStatusTool",
    "get_stack_status",
]